pkg/network/setup/controller.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+	"sync"
+
+	v1 "kubevirt.io/client-go/apis/core/v1"
+	"kubevirt.io/kubevirt/pkg/network/cache"
+)
+
+type Controller struct {
+	setupCompleted    sync.Map
+	ifaceCacheFactory cache.InterfaceCacheFactory
+}
+
+func NewController(ifaceCacheFactory cache.InterfaceCacheFactory) Controller {
+	return Controller{
+		setupCompleted:    sync.Map{},
+		ifaceCacheFactory: ifaceCacheFactory,
+	}
+}
+
+// Setup applies (privilege) network related changes for an existing virt-launcher pod.
+// As the changes are performed in the virt-launcher network namespace, which is relative expensive,
+// an early cache check is performed to avoid executing the same operation again (if the last one completed).
+func (c *Controller) Setup(vmi *v1.VirtualMachineInstance, launcherPid int, doNetNS func(func() error) error, preSetup func() error) error {
```

Comment:
I wonder about this function's signature; In production code, I'm quite sure it will always use `isolationRes.DoNetNS` to tap into net namespaces. Why pass it as an input to the `Setup` func ?

If you're doing this to enable unit testing, can't you set this up in the constructor (having a dedicated constructor for the tests) ?

Reply:
I actually want to decouple the network package from everything else as much as possible. In this case, the network part does not care much about a lot of the isolation functionality, it only cares about running a function in a different netns.

So I hope that I can move out the do-net-ns out of the isolation, and move it to the network package, making this argument gone.
If I am not mistaken, we need only the `pid` in order to detect the NS, and I want the network package to provide this service.

Does this make sense?

Reply:
> So I hope that I can move out the do-net-ns out of the isolation, and move it to the network package, making this argument gone.
If I am not mistaken, we need only the pid in order to detect the NS, and I want the network package to provide this service.

Yes, you only need the PID. OK, while I don't see a reason to spun off another PR with that fix, I'm OK with it.